HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,349 @@
+  desc.add<edm::InputTag>("L2CandTag", edm::InputTag("hltL2MuonCandidates"));
+  desc.add<edm::InputTag>("PreviousCandTag", edm::InputTag(""));
+  desc.add<edm::InputTag>("BeamSpotTag", edm::InputTag("hltOnlineBeamSpot"));
+  desc.add<bool>("MatchToPreviousCand", true);
+  desc.add<double>("MaxDeltaR2L2L3", 0.0025);
+
+  edm::ParameterSetDescription descMuonSelL2;
+  MuonSelectionCuts::fillPSetDescription(descMuonSelL2);
+  desc.add<edm::ParameterSetDescription>("L2CandSelection", descMuonSelL2);
+
+  edm::ParameterSetDescription descMuonSelL3;
+  MuonSelectionCuts::fillPSetDescription(descMuonSelL3);
+  desc.add<edm::ParameterSetDescription>("L3CandSelection", descMuonSelL3);
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+bool HLTMuonL3orL2PreFilter::hltFilter(edm::Event& iEvent,
+                                       const edm::EventSetup& iSetup,
+                                       trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+  if (saveTags()) {
+    filterproduct.addCollectionTag(L3candTag_);
+    filterproduct.addCollectionTag(L2CandTag_);
+  }
+  // get hold of trks
+  auto const L3mucands = iEvent.getHandle(L3candToken_);
+  auto const L2mucands = iEvent.getHandle(L2CandToken_);
+  auto const previousLevelCands = iEvent.getHandle(previousCandToken_);
+  std::vector<reco::RecoChargedCandidateRef> vcands;
+  if (previousLevelCands.isValid()) {
+    previousLevelCands->getObjects(trigger::TriggerMuon, vcands);
+  }
+
+  auto const& beamSpot = iEvent.get(beamspotToken_);
+
+  // Number of objects passing the L3 Trigger:
+  int nL3 = 0;
+  for (unsigned int iMuL3 = 0; iMuL3 < L3mucands->size(); iMuL3++) {
+    reco::RecoChargedCandidateRef L3cand(L3mucands, iMuL3);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L3cand isNonnull " << L3cand.isNonnull();
+    // did this candidate triggered at previous stage.
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L3cand, vcands))
+      continue;
+    // apply all the cuts
+    if (!applySelection(L3cand, beamSpot, cutParsL3_))
+      continue;
+    nL3++;
+    filterproduct.addObject(trigger::TriggerMuon, L3cand);
+  }
+  // get the L3 muons that passed the selection
+  std::vector<reco::RecoChargedCandidateRef> passingL3mucands;
+  filterproduct.getObjects(trigger::TriggerMuon, passingL3mucands);
+
+  // Number of objects passing the L2 Trigger:
+  int nL2 = 0;
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
+    reco::RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L2cand isNonnull " << L2cand.isNonnull();
+    // did this candidate triggered at previous stage.
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L2cand, vcands))
+      continue;
```

Comment:
If I remember well, in the original implementation the "triggeredByPreviousLevel" requirement was applied only on the L3 muons. Does it make sense to apply it to both L2Cands and L3Cands comparing against the same collection (i.e. "previousLevelCands")?